### PR TITLE
Sbovsunovskyi/tdi 42208 support non json documents

### PR DESCRIPTION
**Pull request by slushatel**

### CODE REVIEW

File: couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java [hunk 1]
```diff
@@ -119,74 +68,184 @@ void executeJob(CouchbaseInputConfiguration configuration) {
...
+        assertEquals(arrayStrOriginal, data.get(0).getString("t_array"));
+
+        assertEquals(testData.getColId() + "2", data.get(1).getString("t_string"));
+        log.info("test couchbaseInputDataTest finished");
+    }
 
-        assertNotNull(res);
+    private void insertTestDataToDB(String idPrefix) {
+        Bucket bucket = couchbaseCluster.openBucket(BUCKET_NAME, BUCKET_PASSWORD);
+        // bucket.bucketManager().flush();
```
Comment (Voievidko) [inline couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java]:
```suggestion
```
Remove commented line

File: couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java [hunk 2]
```diff
@@ -119,74 +68,184 @@ void executeJob(CouchbaseInputConfiguration configuration) {
...
+                .put("t_double_max", testData.getColDoubleMax()).put("t_boolean", testData.isColBoolean())
+                .put("t_datetime", testData.getColDateTime().toString()).put("t_array", testData.getColList());
+        return json;
     }
 
     @Test
     @DisplayName("When input data is null, record will be skipped")
     void firstValueIsNullInInputDBTest() {
-        insertTestDataWithNullValueToDB();
+        log.info("test firstValueIsNullInInputDBTest started");
```
Comment (Voievidko) [inline couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java]: I think two comments on each test not needed.

File: couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java [hunk 3]
```diff
@@ -119,74 +68,184 @@ void executeJob(CouchbaseInputConfiguration configuration) {
...
         configurationWithN1ql.setUseN1QLQuery(true);
-        configurationWithN1ql.setQuery("SELECT `t_long_max`, `t_string`, `t_double_max` FROM " + BUCKET_NAME);
+        configurationWithN1ql.setQuery("SELECT `t_long_max`, `t_string`, `t_double_max` FROM `" + BUCKET_NAME
+                + "` where meta().id like \"" + idPrefix + "%\"");
         executeJob(configurationWithN1ql);
 
         final List<Record> res = componentsHandler.getCollectedData(Record.class);
+        assertNotNull(res);
+        // List<Record> data = res.stream().filter(record -> record.getString("id").startsWith(idPrefix))
+        // .sorted(Comparator.comparing(r -> r.getString("id"))).collect(Collectors.toList());
```
Comment (Voievidko) [inline couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java]: Remove commented lines

File: couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java [hunk 4]
```diff
@@ -119,74 +68,184 @@ void executeJob(CouchbaseInputConfiguration configuration) {
     @Test
     @DisplayName("Check input data")
     void couchbaseInputDataTest() {
-        componentsHandler.resetState();
-        insertTestDataToDB();
-
+        log.info("test couchbaseInputDataTest started");
```
Comment (Voievidko) [inline couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java]: remove all logs from test where they exist, or add logs in CouchbaseServiceTest class too

File: couchbase/src/main/java/org/talend/components/couchbase/source/CouchbaseInput.java [hunk 5]
```diff
@@ -78,24 +90,28 @@ public void init() {
 
         columnsSet = new HashSet<>();
 
-        N1qlQueryResult n1qlQueryRows;
+        N1qlQuery n1qlQuery;
         if (configuration.isUseN1QLQuery()) {
-            n1qlQueryRows = bucket.query(N1qlQuery.simple(configuration.getQuery()));
+            /*
+             * should contain "meta().id as `_meta_id_`" field for non-json (binary) documents
```
[the remark this answers is not in the document]
Comment (slushatel) [reply couchbase/src/main/java/org/talend/components/couchbase/source/CouchbaseInput.java]: that is comment for the n1ql part. User should add this field into the query if he wants to read non-json documents

File: couchbase/src/main/java/org/talend/components/couchbase/source/CouchbaseInput.java [hunk 6]
```diff
@@ -105,27 +121,62 @@ private void checkErrors(N1qlQueryResult n1qlQueryRows) {
...
-            return null;
-        } else {
+        // loop to find first document with appropriate type (for non-json documents)
+        while (index.hasNext()) {
             JsonObject jsonObject = index.next().value();
 
-            if (!configuration.isUseN1QLQuery()) {
-                // unwrap JSON (we use SELECT * to retrieve all values. Result will be wrapped with bucket name)
-                jsonObject = (JsonObject) jsonObject.get(configuration.getDataSet().getBucket());
+            // read binary
```
[the remark this answers is not in the document]
Comment (slushatel) [reply couchbase/src/main/java/org/talend/components/couchbase/source/CouchbaseInput.java]: comment is removed as non important

File: couchbase/src/main/java/org/talend/components/couchbase/source/parsers/ParserFactory.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.talend.components.couchbase.source.parsers;
+
+import org.talend.components.couchbase.dataset.DocumentType;
+import org.talend.sdk.component.api.service.record.RecordBuilderFactory;
+
+public class ParserFactory {
+
```
[the remark this answers is not in the document]
Comment (slushatel) [reply couchbase/src/main/java/org/talend/components/couchbase/source/parsers/ParserFactory.java]: private constructor is added

File: couchbase/src/main/java/org/talend/components/couchbase/output/CouchbaseOutput.java [hunk 8]
```diff
@@ -79,6 +92,17 @@ public void release() {
         service.closeConnection(configuration.getDataSet().getDatastore());
     }
 
+    private BinaryDocument toBinaryDocument(String idFieldName, Record record) {
+        ByteBuf toWrite = Unpooled.copiedBuffer(record.getBytes(CONTENT_FIELD_NAME));
+        BinaryDocument doc = BinaryDocument.create(record.getString(idFieldName), toWrite);
+        return doc;
```
[the remark this answers is not in the document]
Comment (slushatel) [reply couchbase/src/main/java/org/talend/components/couchbase/output/CouchbaseOutput.java]: changed

File: couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java (hunk 1 above)
Comment (slushatel) [reply couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java]: deleted

File: couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java (hunk 3 above)
Comment (slushatel) [reply couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java]: done

File: couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java (hunk 4 above)
Comment (slushatel) [reply couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java]: comments are removed

File: couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java (hunk 2 above)
Comment (slushatel) [reply couchbase/src/test/java/org/talend/components/couchbase/source/CouchbaseInputTest.java]: comments are removed

**Pull request merged**